Test/HigherOrderIntrinsicSpecification/ChecksAndUsabilityCoverage.dfy

```diff
@@ -0,0 +1,283 @@
+// RUN: %exits-with 4 %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+// Tests for well-formedness of FunctionCallExpr
+
+function MyFunction(x: int, S: set<object>): real
+  requires x < 100
+  reads S
+{
+  assert true;
+  (x + 12) as real
+}
+
+// In methods, there are no reads checks:
+
+method MCaller0(x: int, S: set<object>) {
+  var myFunctionRequires := MyFunction.requires(x, S);
+  var myFunctionReads := MyFunction.reads(x, S); // error: precondition (x < 100) may not hold
+}
+
+method MCaller1(S: set<object>) {
+  var myFunctionRequires := MyFunction.requires(5, S);
+  var myFunctionReads := MyFunction.reads(5, S);
+}
+
+// In functions, reads checks also matter:
+
+// .requires
+
+function FCallerRequires1(S: set<object>): real {
+  var myFunctionRequires := MyFunction.requires(5, S); // error: insufficient reads clause to call MyFunction.requires
```

Comment:
It seems to as if this proposed axiom `f.requires(t) ==> f.requires.reads(t) == f.reads(t)` is violated here. If that's true, it would make sense to add `assert myFunctionRequires;` (just as below) to make it explicit.

Reply:
This file could overall be enhanced by adding the relevant axioms at the places where errors are thrown.